src/jupyterhub_config/hub_singleuser.py

```diff
@@ -0,0 +1,143 @@
+        env_vars = [f'{key}={value}' for key, value in self.environment.items()]
+
+        cmd = ['sudo', '-E', '-u', self.username, 'env'] + env_vars + [
+            os.path.join(os.environ['JUPYTERHUB_CONFIG_DIR'], 'spawn_notebook.sh')]
+
+        self.log.info(f'Executing command: {" ".join(cmd)}')
+        subprocess.run(cmd, check=True)
+
+    def _ensure_system_user(self):
+        """
+        Create a system user with the given username if it does not already exist.
+        Ensure the group exists before creating the user.
+        Use a file lock to prevent race conditions.
+        """
+
+        lock_file = os.path.join(tempfile.gettempdir(), f'user_creation_{self.username}.lock')
+
+        with open(lock_file, 'w') as lock:
+            fcntl.flock(lock, fcntl.LOCK_EX)
+            try:
+                # Check if user already exists
+                try:
+                    pwd.getpwnam(self.username)
+                    self.log.info(f'User {self.username} already exists')
+                    return
+                except KeyError:
+
+                    # Create the user
+                    self.log.info(f'Creating system user: {self.username}')
+                    useradd_cmd = ['sudo', 'useradd', '-r']
+
+                    if self.groupname:
+                        # Check if the group exists, create if necessary
+                        try:
+                            grp.getgrnam(self.groupname)
+                            self.log.info(f'Group {self.groupname} already exists')
+                        except KeyError:
+                            self.log.info(f'Group {self.groupname} does not exist, creating it.')
+                            subprocess.run(['sudo', 'groupadd', self.groupname], check=True)
+
+                        useradd_cmd.extend(['-g', self.groupname])
+
+                    useradd_cmd.append(self.username)
+
+                    self.log.info(f'Creating system user: {self.username}')
+                    subprocess.run(useradd_cmd, check=True)
+
+            except subprocess.CalledProcessError as e:
+                raise ValueError(f'Failed to create system user: {e}')
+
+            finally:
+                fcntl.flock(lock, fcntl.LOCK_UN)
+
+    def _ensure_workspace_permission(self):
+        """
+        Ensure the user's workspace has the correct permissions.
+        """
+        try:
+            user_info = pwd.getpwnam(self.username)
+        except KeyError:
+            raise ValueError(f'System user {self.username} does not exist')
+        gid = user_info.pw_gid
+        group_name = grp.getgrgid(gid).gr_name
+
+        self.log.info(f'Configuring workspace permissions for {self.username}')
+        subprocess.run(['sudo', 'chown', '-R', f'{self.username}:{group_name}', self.user_dir], check=True)
+        subprocess.run(['sudo', 'chmod', '-R', '777', self.user_dir], check=True)
```

Comment:
Is this still needed or is 700 or 770 ok?

Also this is going to make all files executable, which I assume you don't want

Reply:
👍 changed to 770.

Reply:
Making all the files executable?

Reply:
👍 changed to 750. I am really still trying to make it work first, before considering security, etc.

Reply:
750 still makes the files executable for the owner but doesn't let the group traverse the directories. You'd probably need to do it in 2 steps, one handling directories and one handling files, unless there's some build in way in chmod to do that.

That said if you want to make an issue and move on that's fine

Reply:
I didn't get the concern here. Why we shouldn't make all files executable for the owner?

Reply:
Because most files aren't executable and shouldn't be executed.

Reply:
but the workspace and container is user's. I don't see why not provide max user's need.

Reply:
There is no need to make non-executable files executable

https://security.stackexchange.com/questions/192197/reason-to-not-use-chmod-r-777-on-internal-server-for-project-source-code

Reply:
But it also doesn't hurt. And really not our concerns now.

Reply:
> But it also doesn't hurt.

It does hurt for the reasons in the link and general security principles. But as I said

> That said if you want to make an issue and move on that's fine

Reply:
I really don't know what to put on the issue. Do you think you can make one? I just don't see any security concerns or principles being violated by assigning users 750 permissions on their own workspace within their isolated container.

Reply:
The security concerns are listed above in the SO post

For the ticket, just something like "Don't chmod non-executable files to executable" and link this PR, I'd say

Reply:
👍 https://github.com/kbase/cdm-jupyterhub/issues/96